### PR TITLE
v3

### PR DESCRIPTION
**Pull request by brianlovin**
**Status**
- [x] WIP
- [ ] Ready for review
- [x] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- api
- hyperion (frontend)

**TODO**

- [x] Dynamically hide the nav bar on the community view @mxstbr 
- [x] Add a toggle for communities to enable/disable the chat tab (should create/delete a watercooler thread in the backend)  @mxstbr 
- [x] Render chat messages from watercooler @brianlovin 
- [x] Make automatic pagination of chat tab work @mxstbr 
- [x] Make sending chat messages work @brianlovin 
- [x] Add global community switcher with new navigation @brianlovin 
- [x] Add 'isActive' state to new navigation items from global route @mxstbr 
- [x] Style `isActive` state of navigation tabs @brianlovin 
- [x] Make watercooler opt out for newly created communities (or migrate all existing communities)
- [x] Make thread preview snippets have better formatting - see comments below @mxstbr 
- [x] Notifications icon in side nav needs to be wired up to subscription for badge @mxstbr 
- [x] Mark notifications for threads as seen from thread view @mxstbr 
- [x] DM icon in side nav needs to be wired up to subscription for badge @mxstbr  
- [x] revert message styling changes and tackle in followup PR @brianlovin 
- [x] properly differentiate between "never seen before" and "seen before and something new was added" states in thread feed cards @brianlovin 
- [x] make leave community button way smaller (hide behind gear icon dropdown?) @brianlovin 
- [x] add new dashboard redirect to first/previous community @brianlovin 
- [x] make sure joining/leaving communities updates the global nav @brianlovin 
- [x] change desktop community switch to user command instead of alt @brianlovin 
- [x] figure out where to upsell the desktop app now that there's no dashboard @brianlovin
- [x] Switch to official tippy library https://github.com/CompuIves/codesandbox-client/pull/1584 @brianlovin 
- [x] Implement better thread view sidebar (show trending threads, reduce channel profile size?) @mxstbr  
- [x] Only show threads of channels you are a member of on community view @mxstbr 
- [x] Threads don't get marked as seen in thread feeds when they are opened in a slider @brianlovin  
- [x] Double check z-indexes of dropdowns and modals across the board @mxstbr @brianlovin 
- [x] Store the lastSeen state of each community where a user is a member; use this value to select which community gets loaded in the home view redirect @mxstbr or @brianlovin 
- [x] Explore page loading state
- [x] Show headings in thread snippets, but with normal snippet styling @mxstbr 
- [x] Only show channels a user is a member of in sidebar @brianlovin 
- [x] Upsell channels user is not a member of @brianlovin 
- [x] If you go to a user's profile who you've never messaged before, send a message, then leave the composer modal open. With that second user account, reply to the DM. On the first account, the messages don't appear in the composer modal. If you refresh the user's profile, click 'message' again and open the modal to see your previous messages, the subscription will now work. @mxstbr 
- [x] handle pagination on tall screens @mxstbr - note we did this previously in `components/infiniteScroll/tallViewports` - we can probably re-use this, or if we want a better system maybe we just wrap the `<Loading />` indicators in the infinite scroller with a `react-visibility-sensor` that triggers a `fetchMore` if the element is visible and the component is not currently fetching more...or something like that
- [x] Revert FAB implementation, find a consistent area to prompt for post creation on community/channel views and prompt for DM creation on messages view. @brianlovin 
- [x] clicking on the navigation tab when you have an unread notification doesn't clear the badge immediately @mxstbr 
- [x] community activity indicator in side navigation gets badged even if you are the person who triggered the activity. e.g be viewing a community, publish a new post, the community gets badged
- [x] deleting a thread should return you to that thread's community view, or ideally just to your previous location @brianlovin
- [x] need access to user profile settings on profile view @brianlovin 
- [x] don't use thread modals for now @brianlovin 
- [x] fix background issues when composer modal is open eg on thread view @brianlovin 
- [x] minimize the thread feed refetching loading indicator @brianlovin 
- [x] quick notification styling - ensure full width, remove shadows @brianlovin 
- [x] add way to log out @brianlovin 
- [x] fix participant mention suggestions in threads @brianlovin 
- [x] show unordered and ordered lists in thread snippet, see comment below @mxstbr 
- [x] Backfill `userCommunity.lastSeen` and `community.lastActive` @mxstbr 
- [x] Fix scroll behaviour when clicking into same thread again @mxstbr 
- [x] Add subnav switch between trending and latest threads @mxstbr 
- [x] fix public community feed @mxstbr 
- [x] add link to `/new/community` to sidebar (`+` button at the bottom?)
- [x] rename newly created watercooler chats to "{Community.name} Chat" @mxstbr 
- [x] dont group bot and non-bot messages @mxstbr 
- [x] style subnav switch between trending and latest threads @brianlovin 
- [x] fix titlebar gap on marketing pages @brianlovin 
- [x] signed out navigation @brianlovin
- [x] make sidebar community activity dot real time
- [x] badge the globalTitlebar hamburger icon if either unread messages or notifications @mxstbr
- [x] preserve scroll position when navigating *back* to a thread feed or member feed. eg scroll a long way down a community, click a thread, go back - should be at the same place @mxstbr 
- [x] threads aren't getting marked as read in the threadfeed anymore after they're opened - this seems to always be one "cycle" behind. e.g. open thread 1, go back. thread 1 is marked as unread still. open thread 2. go back. thread 1 is now marked as read, thread 2 is unread still. @mxstbr  
- [x] make sure DM thread list paginates on tall screens @brianlovin 
- [x] make sure list items handle overflow correctly @brianlovin 
- [x] fix ios safari scrolling on views with chatinput @brianlovin 
- [x] fix notifications overflowing viewport @brianlovin 
- [x] fix selecting messages @brianlovin (cc @mxstbr we'll need to redo the ux for this at some point)
- [x] restyle 'load previous' link @brianlovin 
- [x] make sure notifications don't badge if they're for a thread you're currently viewing @mxstbr 
- [x] don't show thread link when embedded @mxstbr 
- [x] set Community.lastSeen on join @mxstbr 

Before ship:
- [x] Flow type all new components @mxstbr @brianlovin 
- [x] make sure desktop app works across all views @brianlovin @mxstbr 
- [x] make sure new user onboarding works
- [x] Fix e2e tests @brianlovin @mxstbr 
- [x] Wrap all new components in error boundaries @brianlovin @mxstbr 

Follow-up:
- [x] Add a way to indicate if there are new things within a community in the sidenav
- [ ] Migrate all existing communities to have a watercooler 
- [ ] cache the top communities for the explore page
- [ ] add switch to show threads that don't have answers
- [ ] Fix thread view flashing with SSR 
- [ ] add usernames to thread feed cards 
- [ ] Add a new visual indicator in local mode that the site is local (eg pink header) 
- [ ] Make search on profile views consistent, try to make tabs a bit more consistent (eg search is on channel/user, but not on community) @brianlovin 
- [ ] handle empty feed state if user is a member of the community, but left all the channels
- [x] refetch the threadfeed query when a user joins/leaves channels 
- [ ] add ability to ban users from community on hover profile
- [ ] rename "watercooler" to "open chat" in notifications @mxstbr 
- [ ] rename all watercoolers to "Open chat" @mxstbr 
- [ ] add 'info' tabs to user profile and channel profile on mobile
- [ ] make quoted messages and blockquotes have same styles
- [ ] make sure user hover profiles don't go beyond the viewport boundaries

### CODE REVIEW

**Comment by brianlovin** [comment]
> Dynamically hide the nav bar on the community view

@mxstbr - can you look into this one specifically? I spent some time trying to get match params into the navbar props but I can't quite figure out how.

**Comment by brianlovin** [comment]
Trying to figure out how to handle scrolling backwards in message history - @mxstbr we definitely need a more global solution to this, I'm not quite sure how to approach it, but we really need to make it so that scrolling backwards in message history just auto paginates backwards while preserving your scroll position after each pagination. 

I'll follow this tutorial tomorrow: https://www.youtube.com/watch?v=nMhIRTfhVWg

**Comment by mxstbr** [comment]
@brianlovin I implemented the dynamic navbar hiding in 29e34b6, what do you think about that solution?

----

I am also not sure how best to do this, I have never had to do that. It should definitely be implemented in the `InfiniteList`, so we can reuse it later.

**Comment by brianlovin** [comment]
> @brianlovin I implemented the dynamic navbar hiding in 29e34b6, what do you think about that solution?

Seems to work, but maybe we could put it in `src/helpers/is-viewing-marketing-page` and just rename that file to something like `should-hide-navbar-based-on-location` or something? Just so we can maintain one list of static marketing pages where we have this behavior.

**Comment by brianlovin** [comment]
> I implemented the dynamic navbar hiding in 29e34b6, what do you think about that solution?

Oh, this breaks when you view a thread or open the composer :P

**Comment by brianlovin** [comment]
@mxstbr I'm adding a to-do list called "Make thread preview snippets have better formatting" - I will need your help with this for sure. Specifically, if we are going to show a snippet we need to make sure to only show paragraph snippets. It's weird to see unformatted code blocks in the snippet, and images at the top of a post just return an empty space.

File: package.json [hunk 1]
```diff
@@ -70,6 +70,7 @@
     "b2a": "^1.0.10",
     "bad-words": "^1.6.4",
     "body-parser": "^1.18.3",
+    "brianlovin-react-tippy": "^1.2.4",
```
Comment (atomiks) [inline package.json]: `react-tippy` is using an ancient version of `tippy.js` from 2017 and is effectively unmaintained -- I recommend switching to [`@tippy.js/react`](https://www.npmjs.com/package/@tippy.js/react) which is actively maintained! CodeSandbox is migrating to it now - see this PR: https://github.com/CompuIves/codesandbox-client/pull/1584

**Comment by mxstbr** [comment]
I am getting lots of warnings here locally that paths do not match due to upper- and lowercasing differences:

```
Module not found: `/Users/mxstbr/projects/withspectrum/spectrum/src/components/SegmentedControl/index.js` does not match the corresponding path on disk `segmentedControl`.
```

Git is not case sensitive by default, did you change the case of those folders? Git would not have picked that up, so we'd have to revert that change.

**Comment by brianlovin** [comment]
Yeah, sorry...I was tired of `importing Foo from './foo'` so I started using uppercase component filenames...to quickly fix this locally just update `SegmentedControl` and `AppViewWrapper`

**Comment by mxstbr** [comment]
That's fine for new components, but let's leave the old ones as-is so git doesn't get confused 👍

**Comment by brianlovin** [comment]
Eh, I'd rather aim for future consistency. In an ideal world we'd iterate towards all Uppercase, but if that's going to be a huge nightmare let's stick with the current camel

File: package.json (hunk 1 above)
Comment (brianlovin) [reply package.json]: Thanks for the followup on this @atomiks! Will try to upgrade.

**Comment by mxstbr** [comment]
@brianlovin f5e640e adds `<Route>`s across the different sidebar nav tabs. I have not implemented any active states as I was not sure at which level/what you wanted them to look like, but essentially in all of those cases it's just a `isActive={!!match}` if that makes sense (`match` is `null` if there is none).

**Comment by brianlovin** [comment]
@mxstbr playing with it now; will this also work for community routes?

**Comment by brianlovin** [comment]
Also, should the sidebar community be "active" if you are viewing a channel view?

**Comment by brianlovin** [comment]
@mxstbr that commit does not work by the way - `match` is `null` in each case

**Comment by mxstbr** [comment]
Yes we'd just wrap each community in a `<Route path={`/${community.slug}`}>{({ match }) => /* ... */}</Route>`!

**Comment by mxstbr** [comment]
No it's not? At least when I visit `/explore` the match returned for that route is the match:

screenshot 2019-03-07 at 17 48 28

**Comment by mxstbr** [comment]
> Also, should the sidebar community be "active" if you are viewing a channel view?

Good question, probably and also on thread views? 😕 Maybe the manual `pathname.match` is the better solution for the community sidebar specifically...

**Comment by brianlovin** [comment]
Hm, I’ll try again but it definitely crashed my app locally and returned null when I console logged…will investigate

Brian

**Comment by mxstbr** [comment]
I just forgot an import 🤦‍♂️ Fixed in the latest commit!

**Comment by mxstbr** [comment]
Latest commit makes it so we automatically create a watercooler thread on community create!

**Comment by brianlovin** [comment]
@mxstbr one thing we need to figure out: if we store the last-viewed community in localstorage, an SSR'd load of the home page will not really work, it will just select the first community in the user's `communityConnection`. To fix this we either need to store the last seen community in the database or don't do anything until the dashboard mounts and then perform a client-side redirect. Any hunch on the best path there?

**Comment by mxstbr** [comment]
I think if anything, we should already store `lastSeen` on all `usersCommunities` records and then select the most recently seen one. That way, adding unread badges to the sidebar will be much easier since we will simply need to track `community.lastActive` and compare those two values, right?

Might as well do that than storing it on the user or something.

**Comment by brianlovin** [comment]
> I think if anything, we should already store lastSeen on all usersCommunities records and then select the most recently seen one. That way, adding unread badges to the sidebar will be much easier since we will simply need to track community.lastActive and compare those two values, right?

Ah, this is a better solution! And there are side effect benefits of all devices being in-sync, and not having to rely on localstorage for one more thing. I'll add this functionality.

**Comment by mxstbr** [comment]
Latest commit fixes and reworks the notifications tab in the side nav. I have also removed all the code related to marking notifications for threads currently being viewed as seen, as that should really just move to the thread view! Going to tackle that next.

**Comment by mxstbr** [comment]
Latest commits sync the notification data to Redux and mark notifications for a thread as seen from the thread container! What do you think about that approach? Imo much much cleaner than our previous hacky `NotificationsTab` implementation.

**Comment by mxstbr** [comment]
Thread view embeds work just fine:
 
Screenshot 2019-03-12 at 11 32 05

**Comment by mxstbr** [comment]
Note to 92e7f02, for some reason the chat tab scrolls on the `main` element, not the `scroller-for-thread-view` element. We should probably change that to scroll on the same element, I just have no clue how to 😅

**Comment by mxstbr** [comment]
Latest commit reworks the query to only show threads of channels you are a member of on the community view. We need to make that clearer by showing which channels you are _not_ a member of in the sidebar, so I've added two new todos above for you @brianlovin. ("Only show channels a user is a member of in sidebar" and "Upsell channels user is not a member of")

**Comment by mxstbr** [comment]
Ugh I hate that piece of code I wrote to handle the scrolling, we need to abstract that better. Going to have more of a think about where best to put that.

**Comment by mxstbr** [comment]
I need to write down all the scenarios so I can get an overview over how the messages should behave across the platform.

**Scroll behaviours**

- Threads: Content above the messages, messages infinitely scroll downwards
- Watercoolers: No content above the messages, messages infinitely scroll upwards—but what if you were linked to a message in the middle, do we infinitely scroll upwards?!
- Direct Message Threads: No content above the messages, infinitely scroll upwards—but what if you were linked to a message in the middle?!

**Loading behaviours**

- Threads: Load first 25 messages, unless you have seen the thread before then only load new messages, unless there is no new messages then load last 25 messages
- Watercoolers: Load last 25 messages
- Direct Message Threads: Load last 25 messages
- Linked directly to a message: Load 25 messages starting from the message you were linked to
- Linked to page (`?msgsafter`/`?msgsbefore`; for SEO although I don't think that even works): Load 25 messages starting from that message

**Comment by brianlovin** [comment]
@atomiks https://github.com/withspectrum/spectrum/pull/4798/commits/efcd115754bb8c85cb1fb584b366417c0c4e64f3 :) - the reason it wasn't working before was because we were on styled components v3, which wasn't forwarding refs. Everything is updated now.

**Comment by brianlovin** [comment]
@mxstbr we should show ul/ol in the thread snippet:
Screenshot 2019-03-14 22 14 58
Screenshot 2019-03-14 22 15 05

**Comment by mxstbr** [comment]
The latest commit makes the thread snippet much smarter by rendering list items AND truncating it after three lines:

Screenshot 2019-03-15 at 09 54 12

**Comment by mxstbr** [comment]
Quick feedback about the thread sidebar: it's strange that the channel profile card has three buttons and also kinda duplicates the CTA from the community profile card.

Screenshot 2019-03-15 at 10 08 18

Same with the logged-out state:

Screenshot 2019-03-15 at 10 08 06

Honestly, let's just get rid of all buttons on the channel profile card, it should just be the channel name, slug and description and you can click on it to go to the channel view, that's it. WDYT?

It's also missing the trending threads which are vital to make sure users don't bounce after coming in from search.

**Comment by mxstbr** [comment]
Latest two commits freaking *nail* the scroll behavior in all threads. 

I also figured out how DMs infinitely scroll—they don't even use the InfiniteScroll component, they just use the NextPageButton, which has a VisibilitySensor that fires when the button becomes visible to fetch the next page!!! 😱 😱 😱

This works super beautifully in threads too, so I switched to that, which means you can now beautifully infinitely paginate in both directions 😍😍😍 

Only things that are missing now is handling message links. Working on it! :soon:

**Comment by mxstbr** [comment]
I cleaned up the new content indicators a bit in the last commit

Screenshot 2019-03-15 at 11 22 32

It is now always clear _why_ a thread is highlighted (either it's a new thread or it has new messages) and the new indicators are consistent and don't hide information (post timestamp vs "New post" indicator).

@brianlovin let me know what you think or feel free to change around some more.

**Comment by mxstbr** [comment]
I went ahead and reworked the thread sidebar a little bit. This is what it looks like now signed in and out:


Screenshot 2019-03-15 at 13 36 09


Screenshot 2019-03-15 at 13 36 28

<br />

@brianlovin do we have an existing pattern for the trending threads list somewhere?

**Comment by mxstbr** [comment]
Made trending threads in the sidebar look a little bit nicer:

Screenshot 2019-03-15 at 13 53 53

Also going to move the desktop app upsell down below the community info and trending threads section!

**Comment by mxstbr** [comment]
Instead of scrolling separately, should the thread sidebar scroll with the rest of the thread view and messages?

**Comment by brianlovin** [comment]
> Instead of scrolling separately, should the thread sidebar scroll with the rest of the thread view and messages?

I don't think so - on long threads you'll never see the sidebar since it will always get scrolled. Same with viewing a profile - if you scroll the feed at all, it's useful to jump directly to channels or initiate one of the community/channel actions like joining/posting.

**Comment by brianlovin** [comment]
@mxstbr added a todo for you above, but if the user is *not* a member of the community, the posts feed should return all public channel posts. Right now if you view a community where you're not a member, no posts show up.

**Comment by mxstbr** [comment]
@brianlovin latest commit adds a subnav to the posts tab in the form of a `select`. Does not look like much right now because I was not sure how you wanted it to look, want to tackle that?

Screenshot 2019-03-18 at 11 17 42

**Comment by mxstbr** [comment]
Added a "Create community" button to the sidenav at the bottom:

Screenshot 2019-03-18 at 11 34 16

**Comment by mxstbr** [comment]
Fixed bot and non-bot messages grouping together, which was very confusing.

**Before**

Screenshot 2019-03-18 at 11 38 25

----

**After**

Screenshot 2019-03-18 at 11 38 13

**Comment by mxstbr** [comment]
Burger badge, done:

Screenshot 2019-03-19 at 09 34 52

**Comment by mxstbr** [comment]
That Peril failure is a fluke, `src/components/tooltip/index.js` has flow enabled! We are back to green apart from the e2e tests 😍

**Comment by mxstbr** [comment]
Latest commit makes it so all embeds replace the link they are embedding. For example:

Screenshot 2019-03-21 at 09 55 55

Turns into

Screenshot 2019-03-21 at 09 56 02

😍

**Pull request merged**